multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,328 @@
+	sort.Slice(gws.Items, func(i, j int) bool {
+		return !gws.Items[i].CreationTimestamp.Before(&gws.Items[j].CreationTimestamp)
+	})
+
+	for _, n := range gws.Items {
+		gatewaysInfo = append(gatewaysInfo, mcsv1alpha1.GatewayInfo{
+			GatewayIP: n.GatewayIP,
+		})
+	}
+
+	// Make sure we only return the last created Gateway for now.
+	return &gatewaysInfo[0], nil
+}
+
+func (r *GatewayReconciler) createorUpdateResourceExport(ctx context.Context, req ctrl.Request,
+	commonArea commonarea.RemoteCommonArea, existingResExport *mcsv1alpha1.ResourceExport, gatewayIP string, creation bool) error {
+	resExportSpec := mcsv1alpha1.ResourceExportSpec{
+		Kind:      common.ClusterInfoKind,
+		ClusterID: r.localClusterID,
+		Name:      r.localClusterID,
+		Namespace: r.namespace,
+	}
+	if creation {
+		resExportSpec.ClusterInfo = &mcsv1alpha1.ClusterInfo{
+			ClusterID:   r.localClusterID,
+			ServiceCIDR: r.serviceCIDR,
+			GatewayInfos: []mcsv1alpha1.GatewayInfo{
+				{
+					GatewayIP: gatewayIP,
+				},
+			},
+		}
+		resExport := &mcsv1alpha1.ResourceExport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: r.leaderNamespace,
+				Name:      r.localClusterID + "-clusterinfo",
+			},
+			Spec: resExportSpec,
+		}
+		resExport.Finalizers = []string{common.ResourceExportFinalizer}
+		if err := commonArea.Create(ctx, resExport, &client.CreateOptions{}); err != nil {
+			return err
+		}
+		klog.InfoS("A ClusterInfo kind of ResourceExport is created", "clusterinfo", klog.KObj(resExport))
+		return nil
+	}
+
+	gwNodeInfo, err := r.getLastCreatedGateway()
+	if err != nil {
+		return err
+	}
+	resExportSpec.ClusterInfo = &mcsv1alpha1.ClusterInfo{
+		ClusterID:    r.localClusterID,
+		ServiceCIDR:  r.serviceCIDR,
+		GatewayInfos: []mcsv1alpha1.GatewayInfo{*gwNodeInfo},
+	}
+	if reflect.DeepEqual(existingResExport.Spec, resExportSpec) {
+		klog.V(2).InfoS("Skip updating ClusterInfo kind of ResourceExport due to no change", "clusterinfo", klog.KObj(existingResExport),
+			"gateway", req.NamespacedName)
+		return nil
+	}
+	klog.V(2).InfoS("Updating ClusterInfo kind of ResourceExport", "clusterinfo", klog.KObj(existingResExport),
+		"gateway", req.NamespacedName)
+	existingResExport.Spec = resExportSpec
+	if err := commonArea.Update(ctx, existingResExport, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&mcsv1alpha1.Gateway{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(r)
+}
+
+// findServiceCIDRByInvalidServiceCreation creates an invalid Service to get returned error, and analyzes
+// the error message to get Service CIDR.
+// TODO: add dual-stack support.
+func (r *GatewayReconciler) findServiceCIDRByInvalidServiceCreation(ctx context.Context) (string, error) {
```

Comment:
The code should be put into a util pkg, and even antrea/pkg/util if it can be used by antrea code too.

@hongliangl, @tnqn  : I remember even AntreaProxy considered to discover Service CIDR?

Reply:
https://github.com/antrea-io/antrea/blob/627239ce6b124b2594ee8e1d37d69d2d476d0845/pkg/agent/route/route_linux.go#L1177

Yes, but it is included in above function.

Reply:
We should refactor and extract shared code.

Reply:
done, move it to controller_utils.go. For now, AntreaProxy still gets the Service CIDR dynamically based on Service ClusterIP, we may refactor this in the future if AntreaProxy plan to use this way to get a ServiceCIDR.